src/dotnet/MLSearch.Service/Indexing/ChatIndexInitializerShard.cs

```diff
@@ -0,0 +1,161 @@
+using ActualChat.Chat;
+using ActualChat.MLSearch.Engine.Indexing;
+
+namespace ActualChat.MLSearch.Indexing;
+
+internal interface IChatIndexInitializerShard
+{
+    ValueTask PostAsync(MLSearch_TriggerChatIndexingCompletion job, CancellationToken cancellationToken);
+    Task UseAsync(CancellationToken cancellationToken);
+}
+
+internal class ChatIndexInitializerShard(
+    IMomentClock clock,
+    ICommander commander,
+    IChatsBackend chats,
+    ICursorStates<ChatIndexInitializerShard.Cursor> cursorStates,
+    ILogger<ChatIndexInitializerShard> log
+) : IChatIndexInitializerShard
+{
+    internal record Cursor(long LastVersion);
+    private const string CursorKey = $"{nameof(ChatIndexInitializer)}.{nameof(Cursor)}";
+    private const int BatchSize = 1000;
+    private readonly Channel<MLSearch_TriggerChatIndexingCompletion> _events =
+        Channel.CreateBounded<MLSearch_TriggerChatIndexingCompletion>(new BoundedChannelOptions(BatchSize) {
+            FullMode = BoundedChannelFullMode.Wait,
+            SingleReader = true,
+            SingleWriter = false,
+        });
+    private long _eventCount;
+    private long _maxVersion;
+    private readonly SemaphoreSlim _semaphore = new(BatchSize, BatchSize);
+    private readonly ConcurrentDictionary<ChatId, (long, long)> _scheduledJobs = new();
+
+    public async ValueTask PostAsync(MLSearch_TriggerChatIndexingCompletion evt, CancellationToken cancellationToken)
+        => await _events.Writer.WriteAsync(evt, cancellationToken).ConfigureAwait(false);
+
+    public async Task UseAsync(CancellationToken cancellationToken)
+    {
+        var cursor = await cursorStates.Load(CursorKey, cancellationToken).ConfigureAwait(false) ?? new(0);
+        _maxVersion = cursor.LastVersion;
+        await Task.WhenAll([
+            ScheduleJobs(cursor.LastVersion, cancellationToken),
+            HandleEvents(cancellationToken),
+            UpdateCursor(cancellationToken),
+        ]).ConfigureAwait(false);
+    }
+
+    private async Task HandleEvents(CancellationToken cancellationToken)
+    {
+        await Task.Yield();
+
+        while (!cancellationToken.IsCancellationRequested) {
+            var evt = await _events.Reader.ReadAsync(cancellationToken).ConfigureAwait(false);
+            Interlocked.Increment(ref _eventCount);
+            if (_scheduledJobs.TryRemove(evt.Id, out var info) && info is var (version, _)) {
+                if (Volatile.Read(ref _maxVersion) < version) {
+                    Volatile.Write(ref _maxVersion, version);
+                }
+                _semaphore.Release();
+            }
+        }
+    }
+
+    private async Task UpdateCursor(CancellationToken cancellationToken)
+    {
+        await Task.Yield();
+
+        var lastEventCount = Volatile.Read(ref _eventCount);
+        while (!cancellationToken.IsCancellationRequested) {
+            try {
+                await clock.Delay(TimeSpan.FromSeconds(30), cancellationToken).ConfigureAwait(false);
```

Comment:
- Not a blocker -
Would be nice to move this into config with a default of 30sec.